### PR TITLE
KAFKA-16842; Fix config validation and support unknown voters

### PR DESCRIPTION
**Pull request by jsancio**
This change fixes the Kafka configuration validation to take into account the reconfiguration changes to configuration and allows KRaft observers to start with an unknown set of voters.

For the Kafka configuration validation the high-level change is that now the user only needs to specify either the controller.quorum.bootstrap.servers property or the controller.quorum.voters property. The other notable change in the configuration is that controller listeners can now be (and should be) specified in advertise.listeners property.

Because Kafka can now be configured without any voters and just the bootstrap servers. The KRaft client needs to allow for an unknown set of voters during the initial startup. This is done by adding the VoterSet#empty set of voters to the KRaftControlRecordStateMachine.

Lastly the RaftClientTestContext type is updated to support this new configuration for KRaft and a test is added to verify that observers can start and send Fetch requests when the voters are unknown.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
@jsancio Sorry making conflicts on this PR. BTW, do you plan to ship (part of/all) KAFKA-17241 into 3.9? I assume all tasks of KAFKA-17241 are included by 4.0 only.

File: raft/src/main/java/org/apache/kafka/raft/internals/VoterSetHistory.java [hunk 1]
```diff
@@ -83,13 +83,9 @@ public Optional<VoterSet> valueAtOrBefore(long offset) {
...
-        Optional<LogHistory.Entry<VoterSet>> result = votersHistory.lastEntry();
-        if (result.isPresent()) {
-            return result.get().value();
-        }
-
-        return staticVoterSet
-            .orElseThrow(() -> new IllegalStateException("No voter set found"));
+        return votersHistory.lastEntry()
+            .map(LogHistory.Entry::value)
+            .orElse(staticVoterSet);
```
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/raft/internals/VoterSetHistory.java:90]: Is there any reason the caller would need to know that they are getting the static voters?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 2]
```diff
@@ -890,47 +890,49 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
...
-        "There must be at least one advertised listener." + (
+        "There must be at least one broker advertised listener." + (
           if (processRoles.contains(ProcessRole.BrokerRole)) s" Perhaps all listeners appear in ${KRaftConfigs.CONTROLLER_LISTENER_NAMES_CONFIG}?" else ""))
     }
     if (processRoles == Set(ProcessRole.BrokerRole)) {
       // KRaft broker-only
-      validateNonEmptyQuorumVotersForKRaft()
+      validateQuorumVotersAndQuorumBootstrapServerForKRaft()
       validateControlPlaneListenerEmptyForKRaft()
-      validateAdvertisedListenersDoesNotContainControllerListenersForKRaftBroker()
```
Comment (mumrah) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:933]: Does removing this allow the broker to advertise the controller listener?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 3]
```diff
@@ -952,32 +954,29 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
       if (controllerListenerNames.size > 1) {
         warn(s"${KRaftConfigs.CONTROLLER_LISTENER_NAMES_CONFIG} has multiple entries; only the first will be used since ${KRaftConfigs.PROCESS_ROLES_CONFIG}=broker: ${controllerListenerNames.asJava}")
       }
-      validateAdvertisedListenersNonEmptyForBroker()
```
Comment (mumrah) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:955]: Don't we still need brokers to advertise their listeners for clients?

File: raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java [hunk 4]
```diff
@@ -279,19 +279,25 @@ private void handleBatch(Batch<?> batch, OptionalLong overrideOffset) {
...
+                        voterSetHistory.addAt(currentOffset, voters);
                     }
                     break;
 
                 case KRAFT_VERSION:
+                    KRaftVersion kraftVersion = KRaftVersion.fromFeatureLevel(
+                        ((KRaftVersionRecord) record.message()).kRaftVersion()
+                    );
+                    logger.info(
+                        "Latest {} is {} at {}",
```
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java]: These will look like: `Lastest kraft.version is 1 at 100`. Can we add "offset" in there to make it clear what the last number is?

**Review by mumrah — COMMENTED** [review]
Thanks @jsancio! A few initial comments

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 3 above)
Comment (mumrah) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:955]: Nvm, I see where we do it below.

**Comment by jsancio** [comment]
> @jsancio Sorry making conflicts on this PR. BTW, do you plan to ship (part of/all) [KAFKA-17241](https://issues.apache.org/jira/browse/KAFKA-17241) into 3.9? I assume all tasks of [KAFKA-17241](https://issues.apache.org/jira/browse/KAFKA-17241) are included by 4.0 only.

@chia7712 I am planning to fix and cherry-pick any sub-task of KAFKA-17241 that has the fix version of 3.9.0. I also tried to keep them at the top of the list. I think there are about 5-6 tasks that we need to complete before 3.9.0.

File: raft/src/main/java/org/apache/kafka/raft/internals/VoterSetHistory.java (hunk 1 above)
Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/internals/VoterSetHistory.java:90]: Yes. In that case they caller would use `lastEntry` which returns an `Optional<LogHistory.Entry<VoterSet>>`.

There are a couple of cases when the leader wants to know only the latest voter set that is in the log/partition and not static voter set.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 2 above)
Comment (jsancio) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:933]: No. There are two advertised listeners that are used by each node when "registering" with the active controller. The brokers use `KafkaConfig#effectiveAdvertisedBrokerListeners()`. The controllers use  `KafkaConfig#effectiveAdvertisedControllerListeners()`. Each of those methods filter the `listeners` and `advertised.listeners` according to the names in `controller.listener.names`.

File: raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java (hunk 4 above)
Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java]: Fixed.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 2 above)
Comment (mumrah) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:933]: Gotcha, so basically anything in `advertised.listeners` that is _not_ present in `controller.listener.names` will be considered a broker listener. Makes sense.

File: raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java [hunk 5]
```diff
@@ -82,7 +82,7 @@ public final class KRaftControlRecordStateMachine {
      * @param logContext the log context
      */
     public KRaftControlRecordStateMachine(
-        Optional<VoterSet> staticVoterSet,
+        VoterSet staticVoterSet,
```
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:85]: I like how you coalesced the "empty" behavior down into VoterSet 👍

**Review by mumrah — APPROVED** [review]
Thanks for the explanations, LGTM pending passing (relevant) tests.

**Pull request merged**